### ISSUE TITLE
Retry delay one kcl error?

### ISSUE THREAD

**Issue by dennybaa** [crossplane-contrib/function-kcl#184]
Hello @Peefy,

I've been learning how to use KCL function. And it's great thanks a lot)! But I've bumped into not very neat behavior, if KCL exits with an error function is immediately reevaluated and this pollutes `info` logs. I wonder if this case is specific to the KCL function only or it's crossplane composite functions architecture feature. Btw I haven't noticed this type of behaviour with patch-and-transform function.

Here's the example:

```
_groupPolicies = option("params").groupPolicies
_spec = option("params").oxr.spec
# Pick vault provider
_provider = "iaas-security-${_spec.env}-vault"
_outputs = option("params").ocds["datasource"]?.Resource?.status?.atProvider?.outputs
_items = []

if _outputs?.group_id:
    # Effectevly wait for a composed resource to appear
    _items = [{
        **_groupPolicies
        metadata.annotations: {"krm.kcl.dev/composition-resource-name" = option("params").name}
        spec.providerConfigRef.name = _provider
        spec.forProvider: {
            policies = _spec.policies
            groupId = _outputs?.group_id
        }
    }]

items = _items
```

To suppress immediate reevaluation and logs pollution I have to manually guard resource creation. Is this expected behavior?

**Comment by dennybaa** [comment]
No, I haven't tried( Let's try to ask @ytsarev . Yury would you mind please sharing your suggestion? What's the best way to handle the delay, should it be inside the KCL function implementation, or maybe the function should return 0 code? The later of course would be odd, however if functions successfully exist crossplane doesn't immediately re-evaluate them, kinda odd mitigation...

Thank you!

**Comment by ytsarev** [comment]
We end up using https://github.com/crossplane-contrib/function-sequencer together with the function-kcl for sequenced creation. Example: https://github.com/upbound/configuration-aws-eks/blob/main/apis/kcl/generate.k#L42-L72

**Comment by Peefy** [comment]
Closing as resolved (workaround documented). The immediate re-evaluation on KCL errors is the standard Crossplane behavior for any composition function — it isn't specific to function-kcl. To delay between steps or guard against premature resource creation:

- Use [@crossplane-contrib/function-sequencer](https://github.com/crossplane-contrib/function-sequencer) to introduce a sequenced delay/wait step (see @ytsarev's example with the EKS configuration).
- Or guard resource emission in the KCL code itself, e.g. `items = _items if _outputs?.group_id else []`.

If you want function-kcl to be able to opt into a longer TTL when it returns conditions like `Ready=False`, please open a follow-up issue (cross-referencing #392). Closing.
